src/gc_roots.cpp

```diff
@@ -93,20 +92,95 @@ class stackInfo : public referenceInfo {
     jmethodID methodId_;
 };
 
+class State {
+public:
+    State() = default;
+
+    State(bool isAlreadyVisited, bool isWeakSoftReachable) : state(0) {
+        setAlreadyVisited(isAlreadyVisited);
+        setWeakSoftReachable(isWeakSoftReachable);
+    }
+
+    void setAlreadyVisited(bool value) {
+        setAttribute(value, 0);
+    }
+
+    void setWeakSoftReachable(bool value) {
+        setAttribute(value, 1u);
+    }
+
+    bool isAlreadyVisited() const {
+        return checkAttribute(0u);
+    }
+
+    bool isWeakSoftReachable() const {
+        return checkAttribute(1u);
+    }
+
+    void updateWeakSoftReachableValue(const State &referrerState) {
+        setWeakSoftReachable(
+                referrerState.isWeakSoftReachable() &&
+                (!isAlreadyVisited() || isWeakSoftReachable())
+        );
```

Comment:
The condition seems too tricky. Can we replace it with something more clear? For example:
```
        if (isWeakSoftReachable()) {
            if (!referrerState.isWeakSoftReachable()) {
                setWeakSoftReachable(false);
            }
        } else {
            if (referrerState.isWeakSoftReachable()) {
                setWeakSoftReachable(true);
            }
        }
```